File: projects/fullstack-image-converter/apps/ios/ConvertMate/Model/ConversionModels.swift
// MARK: - ConversionModels.swift
// Mirrors packages/shared types — platform-native Swift version.

import Foundation
import UniformTypeIdentifiers

// MARK: Formats

enum ImageOutputFormat: String, CaseIterable, Identifiable {
    case jpg, png, webp, heic
    var id: String { rawValue }
    var displayName: String { rawValue.uppercased() }
    var utType: UTType {
        switch self {
        case .jpg:  return .jpeg
        case .png:  return .png
        case .webp: return UTType(mimeType: "image/webp") ?? .data
        case .heic: return .heic
        }
    }
}

enum VideoOutputFormat: String, CaseIterable, Identifiable {
    case mp4, mov, gif
    var id: String { rawValue }
    var displayName: String { rawValue.uppercased() }
}

// MARK: Job

enum JobStatus {
    case pending, processing(progress: Double), done, error(String)

    var isTerminal: Bool {
        switch self { case .done, .error: return true; default: return false }
    }
    var progress: Double {
        switch self {
        case .processing(let p): return p
        case .done: return 1.0
        default: return 0
        }
    }
}

struct ConversionFile: Identifiable {
    let id: UUID
    let name: String
    let size: Int64
    let url: URL          // local file URL (copied to temp dir)
    var inputFormat: String { url.pathExtension.lowercased() }
}

struct ConversionJob: Identifiable {
    let id: UUID
    let file: ConversionFile
    let outputFormat: String    // lowercase ext
    var status: JobStatus = .pending
    var resultURL: URL? = nil

    var outputFileName: String {
        let base = file.name.components(separatedBy: ".").dropLast().joined(separator: ".")
        return "\(base).\(outputFormat)"
    }
}

// MARK: Options

struct ConversionOptions {
    var quality: Double = 0.92   // 0–1
    var keepExif: Bool = true
    var concurrency: Int = 3
}

// MARK: Supported routes (mirrors SUPPORTED_CONVERSIONS in shared)

struct ConversionRoute {
    let from: String
    let to: String
    var label: String { "\(from.uppercased()) → \(to.uppercased())" }
}

let supportedConversions: [ConversionRoute] = [
    .init(from: "webp", to: "jpg"),
    .init(from: "webp", to: "png"),
    .init(from: "png",  to: "jpg"),
    .init(from: "jpg",  to: "png"),
    .init(from: "jpeg", to: "png"),
    .init(from: "heic", to: "jpg"),
    .init(from: "heic", to: "png"),
    .init(from: "avif", to: "jpg"),
    .init(from: "avif", to: "png"),
    .init(from: "mov",  to: "mp4"),
    .init(from: "mp4",  to: "gif"),
    .init(from: "jpg",  to: "pdf"),
    .init(from: "png",  to: "pdf"),
]

func canConvert(from: String, to: String) -> Bool {
    supportedConversions.contains { $0.from == from && $0.to == to }
}

func outputFormats(for inputFormat: String) -> [String] {
    supportedConversions.filter { $0.from == inputFormat }.map(\.to)
}

File: projects/fullstack-media-converter/apps/android/app/src/main/java/com/convertmate/model/ConversionModels.kt
package com.convertmate.model

import android.net.Uri

enum class JobStatus { PENDING, PROCESSING, DONE, ERROR }

data class ConversionFile(
    val id: String,
    val name: String,
    val size: Long,
    val uri: Uri,
)

data class ConversionJob(
    val id: String,
    val file: ConversionFile,
    val inputFormat: String,
    val outputFormat: String,
    val status: JobStatus = JobStatus.PENDING,
    val progress: Int = 0,
    val error: String? = null,
    val resultUri: Uri? = null,
)

data class ConversionOptions(
    val quality: Int = 92,
    val keepExif: Boolean = true,
    val concurrency: Int = 3,
)

data class ConversionRoute(val from: String, val to: String, val label: String)

val SUPPORTED_CONVERSIONS = listOf(
    ConversionRoute("webp", "jpg",  "WebP → JPG"),
    ConversionRoute("webp", "png",  "WebP → PNG"),
    ConversionRoute("png",  "jpg",  "PNG → JPG"),
    ConversionRoute("jpg",  "png",  "JPG → PNG"),
    ConversionRoute("jpeg", "png",  "JPEG → PNG"),
    ConversionRoute("heic", "jpg",  "HEIC → JPG"),
    ConversionRoute("heic", "png",  "HEIC → PNG"),
    ConversionRoute("avif", "jpg",  "AVIF → JPG"),
    ConversionRoute("avif", "png",  "AVIF → PNG"),
    ConversionRoute("mov",  "mp4",  "MOV → MP4"),
    ConversionRoute("mp4",  "gif",  "MP4 → GIF"),
    ConversionRoute("jpg",  "pdf",  "JPG → PDF"),
    ConversionRoute("png",  "pdf",  "PNG → PDF"),
)

fun guessFormat(filename: String): String =
    filename.substringAfterLast('.').lowercase()

fun canConvert(from: String, to: String): Boolean =
    SUPPORTED_CONVERSIONS.any { it.from == from && it.to == to }
